File: web/src/main/java/me/robinvdb/runetracker/database/DataPointDatabaseStub.java
package me.robinvdb.runetracker.database;

import me.robinvdb.runetracker.domain.DataPoint;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;

public class DataPointDatabaseStub implements DataPointDatabase {
    private Map<Integer, DataPoint> dataPoints;
    private final AtomicInteger counter;


    public DataPointDatabaseStub() {
        dataPoints = new HashMap<>();
        counter = new AtomicInteger();
    }

    @Override
    public void add(DataPoint dataPoint) {
        if (dataPoint.getId() == null) {
            dataPoint.setId(counter.getAndIncrement());
        }

        if (dataPoints.containsKey(dataPoint.getId())) {
            throw new DatabaseException("DataPoint with id " + dataPoint.getId() + " already exists");
        }

        dataPoints.put(dataPoint.getId(), dataPoint);
    }

    @Override
    public void addAll(Collection<DataPoint> dataPoints) {
        dataPoints.forEach(this::add);
    }

    @Override
    public void delete(DataPoint dataPoint) {
        dataPoints.remove(dataPoint.getId());
    }

    @Override
    public Optional<DataPoint> get(Integer id) {
        return Optional.ofNullable(dataPoints.get(id));
    }

    @Override
    public Stream<DataPoint> all() {
        return dataPoints.values().stream();
    }
}

File: web/src/main/java/me/robinvdb/runetracker/database/DataPointDatabaseStub.kt
package me.robinvdb.runetracker.database

import me.robinvdb.runetracker.domain.DataPoint

import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.stream.Stream

class DataPointDatabaseStub : DataPointDatabase {
    private val dataPoints: MutableMap<Int, DataPoint>
    private val counter: AtomicInteger

    init {
        dataPoints = HashMap()
        counter = AtomicInteger()
    }

    override fun add(dataPoint: DataPoint) {
        if (dataPoint.id == null) {
            dataPoint.id = counter.getAndIncrement()
        }

        if (dataPoints.containsKey(dataPoint.id!!)) {
            throw DatabaseException("DataPoint with id " + dataPoint.id + " already exists")
        }

        dataPoints[dataPoint.id!!] = dataPoint
    }

    override fun addAll(dataPoints: Collection<DataPoint>) {
        dataPoints.forEach { this.add(it) }
    }

    override fun delete(dataPoint: DataPoint) {
        dataPoints.remove(dataPoint.id)
    }

    override fun get(id: Int?): Optional<DataPoint> {
        return Optional.ofNullable(dataPoints[id])
    }

    override fun all(): Stream<DataPoint> {
        return dataPoints.values.stream()
    }
}
